Client/Frontend/ContentBlocker/ContentBlockerHelper.swift

```diff
@@ -106,42 +106,56 @@ class ContentBlockerHelper {
         return EnabledState(rawValue: pref) ?? .onInPrivateBrowsing
     }
 
-    fileprivate func addActiveRulesToTab() {
-        guard let ruleStore = ruleStore else { return }
+    func addActiveRulesToTab() -> Deferred<[String]> {
+        let result = Deferred<[String]>()
+        guard let ruleStore = ruleStore else { return result }
         let rules = blocklistBasic + (blockingStrengthPref == .strict ? blocklistStrict : [])
         let enabledMode = enabledStatePref
         removeAllFromTab()
 
         func addRules() {
-            for name in rules {
+            let deferreds: [Deferred<String?>] = rules.map { name in
+                let result = Deferred<String?>()
                 ruleStore.lookUpContentRuleList(forIdentifier: name) { rule, error in
-                    self.addToTab(contentRuleList: rule, error: error)
+                    let ok = self.addToTab(contentRuleList: rule, error: error)
+                    result.fill(ok ? name : nil)
                 }
+                return result
+            }
+            all(deferreds).uponQueue(.main) { list in
+                let list = list.flatMap { $0 }
+                result.fill(list)
             }
         }
 
         switch enabledStatePref {
         case .off:
-            return
+            result.fill([])
         case .on:
             addRules()
         case .onInPrivateBrowsing:
             if tab?.isPrivate ?? false {
                 addRules()
+            } else {
+                result.fill([])
             }
         }
+
+        return result
     }
 
     func removeAllFromTab() {
         tab?.webView?.configuration.userContentController.removeAllContentRuleLists()
     }
 
-    fileprivate func addToTab(contentRuleList: WKContentRuleList?, error: Error?) {
+    fileprivate func addToTab(contentRuleList: WKContentRuleList?, error: Error?) -> Bool {
         if let rules = contentRuleList {
```

Comment:
No real need to return true/false here. You can edit this method signature to not include optionals.